File: packages/cli/src/commands/volume.ts
import { confirm } from '../utils/prompts.js';
import { getConfig } from '../utils/config.js';
import {
  backupVolume,
  createBindVolume,
  createDockerVolume,
  listVolumes,
  removeVolume,
  restoreVolume,
  verifyVolume,
} from '../utils/volume.js';
import { log } from '../utils/ui.js';

export async function volumeCreateCommand(options: { bind?: string }): Promise<void> {
  const { volumeName } = getConfig();

  if (options.bind) {
    await createBindVolume(volumeName, options.bind);
    log.success(`Created bind volume ${volumeName} -> ${options.bind}`);
    return;
  }

  await createDockerVolume(volumeName);
  log.success(`Created volume ${volumeName}`);
}

export async function volumeVerifyCommand(): Promise<void> {
  const { volumeName } = getConfig();
  const result = await verifyVolume(volumeName);
  if (result.ok) {
    log.success(`Volume ${volumeName} verified.`);
    return;
  }
  log.error(`Volume verification failed: ${result.reason ?? 'unknown reason'}`);
  process.exit(1);
}

export async function volumeBackupCommand(file: string): Promise<void> {
  const { volumeName } = getConfig();
  await backupVolume(volumeName, file);
  log.success(`Backed up ${volumeName} to ${file}`);
}

export async function volumeRestoreCommand(file: string): Promise<void> {
  const { volumeName } = getConfig();
  await restoreVolume(volumeName, file);
  log.success(`Restored ${volumeName} from ${file}`);
}

export async function volumeDestroyCommand(options: { force?: boolean }): Promise<void> {
  const { volumeName } = getConfig();
  const ok = options.force ? true : await confirm(`Remove volume ${volumeName}?`, false);
  if (!ok) {
    log.warn('Volume removal cancelled.');
    return;
  }

  await removeVolume(volumeName);
  log.success(`Removed ${volumeName}`);
}

export async function volumeListCommand(): Promise<void> {
  const volumes = await listVolumes('dillinger');
  if (volumes.length === 0) {
    log.warn('No dillinger volumes found.');
    return;
  }

  for (const name of volumes) {
    log.plain(name);
  }
}

File: packages/cli_native/src/commands/volume.rs
use anyhow::Result;

use crate::utils::{
    config::get_config,
    prompts::confirm,
    ui::LOG,
    volume::{
        backup_volume, create_bind_volume, create_docker_volume, list_volumes, remove_volume,
        restore_volume, verify_volume,
    },
};

pub async fn volume_create_command(bind: Option<&str>) -> Result<()> {
    let config = get_config();
    let volume_name = &config.volume_name;

    if let Some(host_path) = bind {
        create_bind_volume(volume_name, host_path).await?;
        LOG.success(&format!("Created bind volume {} -> {}", volume_name, host_path));
        return Ok(());
    }

    create_docker_volume(volume_name).await?;
    LOG.success(&format!("Created volume {}", volume_name));
    Ok(())
}

pub async fn volume_verify_command() -> Result<()> {
    let config = get_config();
    let volume_name = &config.volume_name;
    let result = verify_volume(volume_name).await;
    if result.ok {
        LOG.success(&format!("Volume {} verified.", volume_name));
        return Ok(());
    }
    LOG.error(&format!(
        "Volume verification failed: {}",
        result.reason.as_deref().unwrap_or("unknown reason")
    ));
    std::process::exit(1);
}

pub async fn volume_backup_command(file: &str) -> Result<()> {
    let config = get_config();
    let volume_name = &config.volume_name;
    backup_volume(volume_name, file).await?;
    LOG.success(&format!("Backed up {} to {}", volume_name, file));
    Ok(())
}

pub async fn volume_restore_command(file: &str) -> Result<()> {
    let config = get_config();
    let volume_name = &config.volume_name;
    restore_volume(volume_name, file).await?;
    LOG.success(&format!("Restored {} from {}", volume_name, file));
    Ok(())
}

pub async fn volume_destroy_command(force: bool) -> Result<()> {
    let config = get_config();
    let volume_name = &config.volume_name;

    let ok = if force {
        true
    } else {
        confirm(&format!("Remove volume {}?", volume_name), false)?
    };

    if !ok {
        LOG.warn("Volume removal cancelled.");
        return Ok(());
    }

    remove_volume(volume_name).await?;
    LOG.success(&format!("Removed {}", volume_name));
    Ok(())
}

pub async fn volume_list_command() -> Result<()> {
    let volumes = list_volumes(Some("dillinger")).await?;
    if volumes.is_empty() {
        LOG.warn("No dillinger volumes found.");
        return Ok(());
    }
    for name in &volumes {
        LOG.plain(name);
    }
    Ok(())
}
